Switch to pnpm using isolated installs

hassankhan opened the pull request:
# Why

<!--
Please describe the motivation for this PR, and link to relevant GitHub issues, forums posts, or feature requests.
-->

# How

<!--
How did you build this feature or fix this bug and why?
-->

# Test Plan

<!--
Please describe how you tested this change and how a reviewer could reproduce your test, especially if this PR does not include automated tests! If possible, please also provide terminal output and/or screenshots demonstrating your test/reproduction.
-->

# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
-->

- [ ] I added a `changelog.md` entry and rebuilt the package sources according to [this short guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

hassankhan committed:
build: switch from Yarn to pnpm

hassankhan committed:
chore: use `workspace:*` dependency specifiers for `devDependencies`

hassankhan committed:
chore: replace CLI invocations of `yarn` with `pnpm`

hassankhan committed:
ci: update CLI and SDK workflows

kitten committed:
Add missing dependencies

kitten committed:
Remove prepare scripts (yarn semantics for it are wrong)

kitten committed:
create-expo: Remove redundant ts-expect-errors (Readable.fromWeb)

kitten committed:
expo-module-scripts: Fix up pnpm command runs

kitten committed:
expo-module-scripts: Remove default typeRoots configs

kitten committed:
expo-navigation-bar: Fix type import path (self-resolution in TS)

kitten committed:
Switch runPackageScriptAsync from yarn to pnpm

kitten committed:
Install global tools

kitten committed:
Switch to isolated dependencies

kitten committed:
Replace patch-package with pnpm patch

kitten committed:
Add "global" deps (see checkDependenciesAsync for explanation)

kitten committed:
notification-tester: Remove resolution (this never had any effect)

kitten committed:
Hide cycle warning (log-box!)

hassankhan committed:
fix(@expo/cli): fix typecheck errors

hassankhan committed:
fix(@expo/cli): fix type issues in E2E tests

hassankhan committed:
fix: add missing dependencies

hassankhan committed:
fix(@expo/cli): remove `metro-runtime` from `watchFolders` in development projects outside the monorepo

hassankhan committed:
cli,autolinking: add `exports` to `package.json`

kitten commented:
Current Progress:
- #41264 is needed due to self-resolution not working without a `node_modules` structure with `require.resolve` in CommonJS
- #41309 is needed to update Expo Go resolution (otherwise, `expo-av` won't be included)
- #41306 is needed (on most current PRs) to make builds pass

kitten committed:
ci(.github/workflows/cli.yml): Change prepare to build task

kitten committed:
Remove prepare scripts

kitten committed:
Incorporate tools into pnpm workspace

kitten committed:
temp: Add prebuild cli build for log-box and add prepare build

kitten committed:
Fix E2E re-runs by moving atomically before deleting

kitten committed:
Fix `expo/expo` watchFolders exception in CLI overrides

kitten committed:
pnpm dedupe

kitten committed:
Add temp top-level jest-expo workspace dependency

kitten committed:
Fix up various tsc/jest failures

kitten committed:
expo-router: Update snapshots

kitten committed:
eslint-configs: Update snapshots

kitten committed:
local-build-cache-provider,metro-config,expo-router: Rebuild

kitten committed:
router-server: Add missing types/deps

kitten committed:
metro-config: Lint

kitten committed:
local-build-cache-provider: Remove redundant build file

kitten committed:
cli: Update resolve-from mock with explicit error

kitten committed:
Add missing router-e2e deps

kitten committed:
cli/e2e: Fix NODE_PATH pollution for some tests

kitten committed:
Update workflows

kitten committed:
crate-expo: remove semver import and replace

kitten committed:
expo-go: Add missing dependencies

kitten committed:
tools: Fix semver type issue

kitten committed:
cli/e2e: Sanitize RSC payloads

kitten committed:
ncl: Add missing dependencies

kitten committed:
Update install command for create-expo-nightly

kitten committed:
fingerprint: Update snapshots

kitten committed:
bare-expo/macos: Update REACT_NATIVE_PATH and install commands

kitten committed:
bare-expo/e2e/image-comparison: Add missing dependency

kitten committed:
misc: Remove manual `run prepare` from scripts

kitten committed:
bare-expo: Add missing dev dependencies for scripts

kitten committed:
brownfield-tester: Add to workspace with deps

kitten committed:
tools: Update lockfile path in updateWorkspaceProjects

kitten committed:
minimal-tester: Remove redundant deps and babel config

Explicit babel config in pnpm isolated deps requires babel-preset-expo,
but the config is redundant.
